### DISCUSSION TITLE
Update platform: water_heater

### DISCUSSION THREAD

**Discussion by iMicknl** [home-assistant/architecture#518] [Entity Models]
As previously discussed in https://github.com/home-assistant/core/pull/46599, currently the `water_heater` entity doesn't match the code quality and feature completeness of other entities. Especially compared to `climate`, the water_heater is lacking features and has some features only partially implemented.

Hereby a proposal to change the following functionality. I still need to research which existing integrations are impacted by those changes and if we need other changes. 

**Entity services**
Entity services should look at the supported services where needed. Currently all entity services are enabled, but we should look if the service is supported.

**Operation Mode**
Currently operation mode is not required, while HVAC mode is required at climate. Every device should at least have an on/off command, thus I believe operation_mode can be a required attribute. `operation_list` should be changed to `operation_mode`.

If operation modes are hardcoded, we should understand which modes are needed. (and which should be moved to a preset).

**Remove away mode in favour of operation mode**
Not sure if away mode is used widely. It could be also `OPERATION_MODE_OFF` / `OPERATION_MODE_AUTO`.

**Investigate use of `temperature_high` and `temperature_low`.**
Not sure if this is used, it looks like it is currently not used and personally I wouldn't know how this could be used for a water_heater.

Other work which should be done, non breaking;
- Add typing to the entity
- Add target_temperature_step attribute (already supported by front-end)
- STATE_ECO (and others) should be renamed to OPERATION_MODE_*.
- Add async_turn_off functionality with fallback to operation modes, which is already added as a non functional service
- Add async_turn_on functionality with fallback to operation modes, which is already added as a non functional service

@bdraco, @mdz, @joelmoses please chime in :). It would be great if we can all have a look at the current `water_heater` implementations and understand if other features are missing.

**Related**
- https://github.com/home-assistant/core/pull/46522
- https://github.com/simbaja/ha_components/issues/7

**Frontend issues:**
- https://github.com/home-assistant/frontend/issues/5423
- https://github.com/home-assistant/frontend/issues/7225

**Comment by iMicknl** [comment]
## Initial investigation

### 10 core integrations
[atag](https://github.com/home-assistant/core/blob/ac3a6aaa8cdb005b3c900f9d9671e5a859351060/homeassistant/components/atag/water_heater.py)

Simple implementation. Supports setting target temperature, but doesn't set this in `supported_features`. (needs to be changed).
Operation modes used: [STATE_OFF, STATE_ECO, STATE_PERFORMANCE]

[geniushub](https://github.com/home-assistant/core/blob/ac3a6aaa8cdb005b3c900f9d9671e5a859351060/homeassistant/components/geniushub/water_heater.py)

Supports setting target temperature.
Operation modes used: STATE_OFF, STATE_AUTO, STATE_MANUAL

[hive](https://github.com/home-assistant/core/blob/3381e2f65a0cd5e8e4456bdc21d4703560445b7b/homeassistant/components/hive/water_heater.py)

Operation modes used: STATE_ECO, STATE_ON, STATE_OFF
Implements async_turn_on and async_turn_off

[incomfort](https://github.com/home-assistant/core/blob/ac3a6aaa8cdb005b3c900f9d9671e5a859351060/homeassistant/components/incomfort/water_heater.py)

Doesn't support any commands? Can show the current temperature / operation_mode, but no interaction possible.

Sets `mdi:thermometer-line` as the icon
Sets attrs `display_code", "display_text", "is_burning`

[melcloud](https://github.com/home-assistant/core/blob/ac3a6aaa8cdb005b3c900f9d9671e5a859351060/homeassistant/components/melcloud/water_heater.py)

Operation modes used: unknown, they come directly from pymelcloud package
Supports setting target temperature.
Implements async_turn_on and async_turn_off

[wink](https://github.com/home-assistant/core/blob/ac3a6aaa8cdb005b3c900f9d9671e5a859351060/homeassistant/components/wink/water_heater.py)

Operation modes used: STATE_ECO, STATE_ELECTRIC, STATE_GAS, STATE_HEAT_PUMP, STATE_HIGH_DEMAND,  STATE_OFF, STATE_PERFORMANCE,
Supports setting target temperature.
Supports away mode (mapped to vacation mode)
Sets attrs `vacation_mode`, `rheem_type`


[vicare](https://github.com/home-assistant/core/blob/ac3a6aaa8cdb005b3c900f9d9671e5a859351060/homeassistant/components/vicare/water_heater.py)

Operation modes used: OPERATION_MODE_ON = "on", OPERATION_MODE_OFF = "off" (self-defined)
Supports setting target temperature.
Precision is set to whole, but target_temperature_step is not set.

[evohome](https://github.com/home-assistant/core/blob/ac3a6aaa8cdb005b3c900f9d9671e5a859351060/homeassistant/components/evohome/water_heater.py)

Operation modes used: STATE_AUTO, STATE_ON, STATE_OFF
Supports away mode (mapped to off mode)
Sets attrs `"dhwId", "activeFaults", "stateStatus", "temperatureStatus"` (but current_temp not set)
Sets icon to `mdi:thermometer-lines`
Precision is set to whole for some devices, but target_temperature_step is not set.

[econet](https://github.com/home-assistant/core/blob/260d9f8e16cf8593a5629a660f0864209afbbcfd/homeassistant/components/econet/water_heater.py)

Operation modes used: STATE_ECO, STATE_ELECTRIC, STATE_GAS, STATE_HEAT_PUMP, STATE_HIGH_DEMAND,  STATE_OFF, STATE_PERFORMANCE,
Supports setting target temperature.
Supports away mode for some devices (mapped to ??)

[tado](https://github.com/home-assistant/core/blob/50e11773eeb9e1c8b803c3515a27db705eefb23a/homeassistant/components/tado/water_heater.py)

Operation modes used: MODE_AUTO = "auto", MODE_HEAT = "heat", MODE_OFF = "off"
Supports setting target temperature for some devices.
Adds service `set_water_heater_timer`

### Custom integrations:

[somfy tahoma](https://github.com/iMicknl/ha-tahoma/blob/master/custom_components/tahoma/water_heater.py)

Multiple devices (boilers) are supported in TaHoma implementation, all with a different feature set. One of them has an 'absence mode' which is currently mapped to `away` mode. One of them is using `target_temperature_high` and `target_temperature_low`, but this is untested and looks non functional.

[nibe](https://github.com/elupus/hass_nibe/blob/master/water_heater.py)

**Reply by elupus** [reply]
https://github.com/elupus/hass_nibe/blob/master/water_heater.py

**Reply by elupus** [reply]
nibe_heatpumps: https://github.com/home-assistant/core/pull/79099

**Reply by elupus** [reply]
target_temperature_high/low are useful since they represent a start heating, stop heating range and are nice to have visualized in state history.

**Comment by mdz** [comment]
Thanks for working on this!

> **Operation Mode**
> Currently operation mode is not required, while HVAC mode is required at climate. Every device should at least have an on/off command, thus I believe operation_mode can be a required attribute. `operation_list` should be changed to `operation_mode`.

Notably, SmartTub does not have an on/off. Hot tubs generally are always-on when power is supplied. Do you think it makes sense to account for this in water heater?

(It does have an operation mode though)

**Reply by iMicknl** [reply]
Is this a custom component or did I miss this one? Which operation modes does it support?

I didn’t think that a hot tub is also a usecase for the water_heater entity! Good addition.

**Reply by mdz** [reply]
It was just merged into dev a few days ago. It currently uses the climate platform, primarily because I didn't know that water heater existed. But once I noticed it, I found that it doesn't seem to have an equivalent of `hvac_action` to indicate when the heater is active. That's very useful for monitoring energy consumption in a hot tub scenario. That would be a good thing to add.

**Reply by iMicknl** [reply]
@mdz I believe the `operation_mode` is the equivalent of the `hvac_action`.

**Reply by mdz** [reply]
`operation_mode` seems like an analog of `hvac_mode`, i.e. a mode which is set by the user:

```
HVAC_MODE_OFF = "off"                                                                                                  
HVAC_MODE_HEAT = "heat"                                                                                                
HVAC_MODE_COOL = "cool"                                                                                                
# The device supports heating/cooling to a range                                                                       
HVAC_MODE_HEAT_COOL = "heat_cool"                                                                                      
```

`hvac_action` is the dynamic state of the unit, without any intervention by the user:

```
# This are support current states of HVAC                                                                              
CURRENT_HVAC_OFF = "off"                                                                                               
CURRENT_HVAC_HEAT = "heating"                                                                                          
CURRENT_HVAC_COOL = "cooling"                                                                                          
CURRENT_HVAC_DRY = "drying"                                                                                            
CURRENT_HVAC_IDLE = "idle"                                                                                             
CURRENT_HVAC_FAN = "fan"                                                                                               
```

**Reply by elupus** [reply]
I just opened a pull request to disentangle these two attributes (mode and current): https://github.com/home-assistant/core/pull/78423

**Comment by dgomes** [comment]
I've got a pending proposal for a generic water_heater https://github.com/home-assistant/core/pull/44605

**Reply by frenck** [reply]
Different and unrelated discussion: IMHO. Also unrelated: I don't think we should add more "generic" platforms. If anything, we should get rid of those.

**Comment by iMicknl** [comment]
Today I found out that presets aren't supported yet for the water heater yet.. Would need to be added as well.

**Comment by MatsNl** [comment]
Regarding the water heater in the Atag integration, this is actually suboptimal as well. 
The device can be in Eco and Comfort modes (comfort i mapped to Performance, but really it should be Comfort). Then aside from these two modes, it can be either active (on/heating) or not (off).

Thus, in my opinion the operating mode and current status should be decoupled, similar to how this is done for climate devices with hvac mode and hvac action

**Comment by BradleyFord** [comment]
This is also similar to the humidifiers

**Comment by farmio** [comment]
Support for a "legionella protection" operation mode would also be nice.

**Comment by davet2001** [comment]
Adding to the discussion here rather than creating a new one.
The [hive integration has a boost mode but this is not visible to HA](https://github.com/home-assistant/core/issues/58795) because of water_heater's current limitations.
This creates the strange situation where the water tank heater is reported as off while the boost is active.  

I feel that if it can be set to ON, OFF, ECO already, then BOOST should be allowed too.

So my short term suggestion would be to [add a boost state to the generic water heater (branch)](https://github.com/davet2001/home-assistant/tree/hive_boost_status).

I agree that it could be better aligned to climate.  Any objections to my short term suggestion?

**Comment by davet2001** [comment]
Trying to think of a good first step.  Could we refactor to add `PRESET_MODE` exactly the way `climate` currently does it?  If that is agreed, then we can make a start.  I think it should be ok to move to a pattern used elsewhere in the project, and it should be quite quick to get agreement (e.g. "yes that works, please use it for water_heater too").

**Comment by klacol** [comment]
Hi, good to see the evolvement of the platform waterheater. I have worked on an integration of a CLAGE continuous water heater the last weeks:

https://github.com/klacol/homeassistant-clage_homeserver

This is an device that works not like a storage or tank. It heats the water up on the fly. I propose to differentiate the type of waterheater somehow.

Type 1: `Hot water tank`
Type 2: `Continuous water heater`

Both types share many entities, e.g.
- TemperatureIn
- TemperatureOut
- TargetTemperature
- WaterConsumption

But both types also have also different entities, related to their type.

I just wanted to share that aspects to this good discussion here.

**Reply by elupus** [reply]
Continuous water heater would be interesting for outgoing hotwater to radiators. It would be interesting to be able to map temperature curve offsets/curve angle up for modification and display. But maybe w should keep that separate.

**Reply by elupus** [reply]
*Additions*
* `operation_mode`:
  * [..., FLOOR_DRYING]
* `current_operation`:
  * [..., DRYING_FLOORS]
* `current_temperature_in`
  * Temperature of fluid coming back to heater
* `current_temperature_out`
  * Temperature of fluid leaving or about to leave heater. This is not the same as current_temperature, since that is the charging temperature of the heater, used for controlling availability of hotwater. 
* `flow_rate`
  * Flow rate of fluid leaving heater
* `target_temperature_offset` `SUPPORT_TEMPERATURE_OFFSET`
  * Wanted offset from a calculated temperature setpoint, used for radiator/floor heating systems
* `device_class`:
  *  [..., HEATING]

**Comment by elupus** [comment]
I would like to propose the following.

*Changes*
 * `current_operation`:
   * CurrentOperation.[None, IDLE, HEATING, ~~LEGIONELLA_PREVENTING~~, ~~BOOSTING~~],
   * None should be used if heater doesnt know if its currently heating or not.
 * `operation_mode` (`operation_modes`):
   * OperationMode.[OFF, ON, ~~BOOST~~]
   * ON, OFF, ~~short term activation where heater normally normally return back to ON after some delay.~~
   * Operation mode of a pump does not need to be updated, to match current operation (nor include) if the activity is a background process. Only if the user have a option of cancelling/starting the background operation should they be included in operation_modes.
 * `state`:
   *  Will be mapped to operation_mode instead of current_operation. This matches how climate entities behave.

*Additions*
 * `preset_mode` (`preset_modes`)
   * PresetMode.[ECO, NORMAL, LUXARY, AUTO, FROST_PROTECTION, BOOST, AWAY]
   * Preset modes are to be allowed to be customized per device_class/integration so it doesnt need to be exhaustive. Just the most common so we get translations.
   * `SUPPORT_PRESET_MODE` controls ability to set presets.
* `device_class`:
  *  [domestic_hot_water, pool]

*Unchanged*
 * `target_temperature_high`, `target_temperature_low`:
   * Kept as is to allow for display of setpoints in history graph
 * `target_temperature`:
   * Kept in line with climate, but probably should like to see it dropped in favor of setting high and low equal
 * `current_temperature`:
   * Kept to represent current water temperature
 * `SUPPORT_TARGET_TEMPERATURE` controls ability to set a target, not the ability to display one
 * `SUPPORT_OPERATION_MODE` controls ability to set a operation mode


*Removed*
  * `operation_list` removed in favor of `operation_modes` to line up better with name of mode.
  * GAS, HEATPUMP, ... types of states. 
  * Separate away mode, compatibility shim for services initially?

**Reply by elupus** [reply]
Possible add a boost_time, boost_times to indicate for how long a boost interval lasts. Nibe supports single increase, 1h, 3h, 6h. Alternatively we allow that as part of operation mode. But we can think about that later I think. Lets avoid making this large for now.

**Reply by elupus** [reply]
Is BOOST a good name? nibe call it temporary luxuary.

**Reply by elupus** [reply]
I decided to start an implementation. let's see if i'll regret it again: https://github.com/home-assistant/core/pull/81913

**Reply by elupus** [reply]
@iMicknl , @davet2001 can I get a comment and perhaps a +1

**Reply by iMicknl** [reply]
@elupus I will try to have a more extensive look this weekend, but @nyroDev (who is currently porting the water heaters to core), already gave some feedback on your PR.

**Reply by elupus** [reply]
FROST_PROTECTION was suggested. Thinking about it, it seems most logical as a PRESET. Its a a setpoint of very low heating temperatures. 

Ps. In general presets are going to be allowed to be customizable per integration.

**Reply by davet2001** [reply]
@elupus Thanks for starting this. 

Seems like a good initiative and I'll look at the PR as soon as I get an opportunity.👍

**Reply by nyroDev** [reply]
> FROST_PROTECTION was suggested. Thinking about it, it seems most logical as a PRESET. Its a a setpoint of very low heating temperatures.
> 
> Ps. In general presets are going to be allowed to be customizable per integration.

In this case, Boost should also be a preset : it's a setpoint of very high heating.

I also think many water heater do have a mode that you can set `physically` on it to frost_protection.  
That's why I think it should be a mode and not a preset: i exists. (I also think it should be added to climate too, but that's another topic)

**Reply by nyroDev** [reply]
Actually, after rereading the climate entity, you're right, FROST_PROTECTION should be a preset.  
That's also the case for Boost.

**Reply by elupus** [reply]
Boost is a temporary thing thou. It will return to a normal state when done. If its a preset, the normal preset is not know. Its more a temporary override of the preset.

**Reply by elupus** [reply]
Away also a temporary thing, however not something the pump ends automatically. Maybe we should let that be a preset too.

**Reply by nyroDev** [reply]
ReReRe reading and thinking again, and reading the climate documentation too.

Regarding `current_operation` and `operation_mode`, I think their names should match the climate device.  
In this case:  
- `operation_mode` could be renamed to `wh_mode` and will represent the current mode of the water heater, meaning how the user actually set it's water heater to work. If integration need a new mode, it should use a preset (like climate) It represent a kind of long term configuration of the device. possibles values could be `ON`, `OFF`, `FROST_PROTECTION`, `AUTO`, `MANUAL`
- `current_operation` could be renamed to `wh_action` and will represent what the water heater is currently doing (only values could be `HEATING` or `None` I think, Water heater could not do anything else)
- `preset_mode` will represent the current user configuration for this device. Integration could add more preset. it could represent a long term preset or a preset mode that will expires by itself. Regular values could be `ECO`, `BOOST`, `COMFORT`,  `AWAY`,  `LEGIONNALA_PROTECTION`

I'm not 100% sure about `LEGIONNALA_PROTECTION` as I didn't faced a device with it.  
As I can tell from Google, it consists of boiling the water to the upper value possible for a certain amount of time.  
The definition above match it.

Regarding `FROST_PROTECTION`, it does make sense to set it in wh_mode : When having a water heater in an house not used, this mode will be used without any preset mode.
It could also be set as `AWAY`: it will less verbose, but more versatile and could be used differently by integration.  
In that case, preset_mode should add `FROST_PROTECTION`

It will not represent 100% of the functionnality I saw.

For instance, in some device you could set a boost mode for a certain amount of time.
Maybe we could add a new supported_features with `BOOST_DURATION` and will require a new function `async_boost_duration(int)` 

@davet2001 regarding your comment on the PR :  

> In the case of my hive hot water controller, during boost there is no way of knowing whether it is heating or not, nor the temperature nor the target temperature, only that it is in boost mode (ultimately its output to the gas boiler is one on/off signal and only the boiler (not hive or HA) knows whether it's actually heating.).

According to my definition above, when in BOOST mode :  
- `wh_mode` will be `MANUAL`
- `wh_action` could be heating or None, depending of what you think is the most usufuel (I guess heating could be ok)
- `preset_mode` will be `BOOST`

What do you think?

**Reply by elupus** [reply]
Firstly. I want to get something done. I want a conclusion, not really a discussion on naming. I prefer to keep names as close to current as possible. I dont see a benefit of renaming fields, it causes more work everywhere 

If we want attribute names to be globally unique, i see a point. But as far as i know we dont do that.

Leginella protection is not a preset. Its not something you set the heater to for a long period of time.

Boost is temporary. It need to be able to return to a normal state without user interaction, and user should know what it returns to. What is normal after manual and boost?

Away could equate frost protection, agree. I think both are presets in mode ON.

Auto/Manual in mode i have a hard time with. This is not a representation of the heating element. Its always a temperature controlled feature. What is auto/manual, is it just lack of temp settings?

Boost duration is interesting. But lets leave that out for now. I want to leave continous heters out too for now.

**Reply by nyroDev** [reply]
> Firstly. I want to get something done. I want a conclusion, not really a discussion on naming. I prefer to keep names as close to current as possible. I dont see a benefit of renaming fields, it causes more work everywhere

You're changing the available values, so work will have to be done anyway everywhere I guess.

> Leginella protection is not a preset. Its not something you set the heater to for a long period of time.

I'm not sure to follow you on this one.  
How do you intend to support it? (if you want to)
 
> Boost is temporary. It need to be able to return to a normal state without user interaction, and user should know what it returns to. What is normal after manual and boost?

I guess that will be the role of integration to decide it, by updating the preset when it has too.
And it will depend: some will come back to ON, some to AUTO.
In my mind, it will basically come back to what it was before setting the boost.

> Away could equate frost protection, agree. I think both are presets in mode ON.

IMHO, Away and ON are not the same thing and could not be compatible.
ON means regular function, provide heat water everytime.
AWAY means keep the least temperature needed to not have problem with the installation.
Could you list what you intend to provide for Mode and preset please? Because I'm not sure to follow what you intend to have for available modes.
I think it will be clearer to understand

> Auto/Manual in mode i have a hard time with. This is not a representation of the heating element. Its always a temperature controlled feature. What is auto/manual, is it just lack of temp settings?

Some Water heater do have functionnality to set hours to provide water heater, and hours when hot water will not be needed.  
In that case, AUTO could be used to indicates: "The water heater is functionning using a programming"

Manual is similar to boost, but not exactly the same: the user could set a custom target temperature manually, different from the regular target temperature and disable the AUTO mode.

It's also mean when the preset is set to boost, I think most integration will also set mode to MANUAL and then go back to what it was before.

> Boost duration is interesting. But lets leave that out for now. I want to leave continous heters out too for now.

Looks great to me, the amount of work for this change is already big!

**Reply by elupus** [reply]
> 
> > Leginella protection is not a preset. Its not something you set the heater to for a long period of time.
> 
> I'm not sure to follow you on this one. How do you intend to support it? (if you want to)

The pump periodically runs legionella program by itself. When it's doing so, it will be set to:

```yaml
mode: ON
current: LEGIONELLA_PROTECTION
preset:  anything goes
```
 
If user explicitly requested a legionella run (and the pump support such a request), or user can explicitly turn of a legionella run.

```yaml
mode: LEGIONELLA_PROTECTION
current: LEGIONELLA_PROTECTION
preset: ECO (or anything)
```

Once complete:

```yaml
mode: ON
current: IDLE or HEATING
preset: ECO (or anything)
```


> > Away could equate frost protection, agree. I think both are presets in mode ON.
> 
> IMHO, Away and ON are not the same thing and could not be compatible. ON means regular function, provide heat water everytime. AWAY means keep the least temperature needed to not have problem with the installation. Could you list what you intend to provide for Mode and preset please? Because I'm not sure to follow what you intend to have for available modes. I think it will be clearer to understand

Normally pump will be in:

```yaml
mode: ON
current: IDLE or HEATING or LEGIONELLA_PROTECTION
preset: anything but AWAY
```

You switch pump to away. It stays in away indefinitely, until preset is manually returned to ECO, NORMAL, SMART, AUTO.....

```yaml
mode: ON
current: IDLE or HEATING or LEGIONELLA_PROTECTION
preset: AWAY
```
 
> > Auto/Manual in mode i have a hard time with. This is not a representation of the heating element. Its always a temperature controlled feature. What is auto/manual, is it just lack of temp settings?
> 
> Some Water heater do have functionnality to set hours to provide water heater, and hours when hot water will not be needed. In that case, AUTO could be used to indicates: "The water heater is functionning using a programming"

That would be a AUTO preset. You can have AUTO_ECO, AUTO_COMFORT whatever the specific integration needs.

```yaml
mode: ON
current: IDLE or HEATING or LEGIONELLA_PROTECTION
preset: AUTO
target_temperature_high: None
target_temperature_low: None
target_temperature: None 
```

> Manual is similar to boost, but not exactly the same: the user could set a custom target temperature manually, different from the regular target temperature and disable the AUTO mode.

```yaml
mode: ON
current: IDLE or HEATING or LEGIONELLA_PROTECTION
preset: ECO, NORMAL, ....
target_temperature_high: None
target_temperature_low: None
target_temperature: 22.0
```

In general above is easy to understand i think, and should be easy to implement.

**Reply by nyroDev** [reply]
@elupus Thanks for all your explanation, I took time to read and confront with what I already saw.
Everything looks clearer and makes sense.  

I think I don't have anything else to say, thank you for the discussion here (and sorry for bothering you, I hope the discussion helped a little bit)

**Reply by frenck** [reply]
Legionella protection sounds like something that doesn't fit here. I don't see any integration that supports it. Unless you can provide multiple integrations that could support this, we should not add it.

**Reply by elupus** [reply]
I know nibe heatpump support it, i've already removed it as a Mode, i think we can keep a translation for preset for it.

**Reply by elupus** [reply]
What i meant was, I think it would be good to keep as current operation. It is a VERY common thing for any water heater in domestic use since keeping water at semi-high temp is bound to grow bacteria. It is also a costly operation to perform so being able to indicate it's active make sense.

I will check if i can find others that can indicate it thou.

**Reply by elupus** [reply]
I've marked boost, and legionella protection as removed in proposal and added boost to presets. So this means it's a entity that is almost always on (since i don't think many DHW systems is ever turned off) that now have presets.

**Reply by geduxas** [reply]
I am here a little bit late, but i am missing some features for water heater entity. To use it for heat pump where is need some more mode buttons.

Force DHW to tell heatpump that i need to reheat tank.
Enable power booster - To tell pump to use additional resistor for heating
also would be nice to have heating/idling states (and button) as in climate, for history graphs (and maybe reuse to force reheating), as state_on/state_off could be used to enable/disable DHW heating.

truly don't know how to use STATE_HEATPUMP at all..

**Comment by elupus** [comment]
It was mentioned that some heater dont know if its heating or not. In that case i think current_operation should be set to None. They can still set it to boost when they actually know its in boost mode.

**Reply by frenck** [reply]
Boost modes in climate are generally considered presets. I'm not sure why that would be implemented differently in case of a water heater.

**Reply by elupus** [reply]
I've tried to explain it above. When triggering a boost mode, it make sense to know what the pump returns to. Will it go back to eco, luxury, or what? So i think it make sense to keep that as a mode, then preset remains static things that doesn't automatically change around. Mode will be a reaction to a user request that is expected to return back to ON.

**Reply by frenck** [reply]
Sorry, I think it still not makes sense compared to our climate entity and will thus create an inconsistency. Therefore I would no opt to add it.

**Reply by elupus** [reply]
Do you object to it in both current and in mode? I can live with it being removed from mode, but i would like it to be indicated in graphs for what the pump is actually doing, same as legionella protection.

Ie.. Answer to why the heck did the heater heat my water to 75 degrees suddenly last night, i have it ECO. Presets are not usable for that.

**Reply by frenck** [reply]
> Do you object to it in both current and in mode? 

Yes, that would be highly confusing.

**Reply by elupus** [reply]
Why. Current have little to do with Mode. Mode is a request, Current is what it is doing. That is why we have a separation.

A climate can be in heat_cool, mode but in idle action.

**Reply by elupus** [reply]
But if it moves us forward, then fine. Then everything moves to presets and its an on/off entity with just off/idle/heating.

**Comment by Jc2k** [comment]
One thing thats come up in other platform tickets: Reducing the number of entity attributes where it makes sense to break them out as separate sensors. I've heard this come up in the context of climate - iirc regrets by core team members that e.g. current_temperature and hvac_action weren't made separate sensors in climate v2. This would create an inconsistency with climate of course, but as it's something I've seen something discussed elsewhere I thought I should mention it.

**Reply by elupus** [reply]
Then lets drop the entity and create something that can visualize temperature and current action like climate can without support for any modes or temperature settings.

The current status quo of water heater is worse than anything. (Its even broken in frontend at the moment since its half using climate stuff)
